lib/model/compose.dart

```diff
@@ -0,0 +1,103 @@
+  // wouldn't notice the internal fence that we're steering clear of,
+  // and if servers *do* start following the spec by noticing indented internal
+  // fences, then this client behavior will be nice.
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*$';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest opening backtick fence that's longer than any in [content].
+///
+/// Expressed as a number of backticks.
+///
+/// Use this for quote-and-reply or anything else that requires wrapping
+/// Markdown in a backtick fence.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedOpeningBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length + 1);
+  }
+  return result;
+}
+
+/// Wrap Markdown [content] with opening and closing backtick fences.
+///
+/// For example, for this Markdown:
+///
+/// ```javascript
+/// console.log('Hello world!');
+/// ```
+///
+/// this function, with `infoString: 'quote'`, gives
+///
+/// ````quote
+/// ```javascript
+/// console.log('Hello world!');
+/// ```
+/// ````
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+// In [content], indented code blocks
+//   ( https://spec.commonmark.org/0.30/#indented-code-blocks )
+// and code blocks fenced with tildes should make no difference to the
+// backtick fences we choose here; this function ignores them.
+String wrapWithBacktickFence({required String content, String? infoString}) {
+  assert(infoString == null || !infoString.contains('`'));
+  assert(infoString == null || infoString.trim() == infoString);
+
+  StringBuffer resultBuffer = StringBuffer();
+
+  // (A) Why not panic on dangling opening fences
+  //     (ones without a corresponding closing fence)?
+  //     Because the spec allows leaving the closing fence implicit:
+  //     > If the end of the containing block (or document) is reached
+  //     > and no closing code fence has been found,
+  //     > the code block contains all of the lines after the opening code fence
+  //     > until the end of the containing block (or document).
+  //
+  // (B) Why not look for dangling closing fences (ones without an opening fence)?
+  //     Because technically there's no such thing:
+  //     they would be indistinguishable from dangling opening fences,
+  //     and parsers (and this function) will treat them that way.
+  final fenceLength = getUnusedOpeningBacktickFenceLength(content);
```

Comment:
To your comment [above](https://github.com/zulip/zulip-flutter/pull/179#issuecomment-1590237155) about this comment: I think a good direction to go in commenting on this would be on the lines of @andersk's comment at https://github.com/zulip/zulip-flutter/pull/179#discussion_r1228712591 , after making the simplification I point out in the comment before this one.

Or perhaps just take the two "what we need to do" points from the end of that comment of mine, with a link to that comment of Anders's.